File: fragments/solana_rpc/solana_rpc_utils.test.ts
import { initRpcClient, initRpcSubscriptionsClient, waitForSlot } from "./solana_rpc_utils";
import assert from "node:assert/strict";
import test, { describe } from "node:test";
import process from "node:process";

describe("solana rpc utils", () => {
  test("initRpcClient", async () => {
    const client = initRpcClient();
    const blockHeight = await client.getBlockHeight().send();
    assert.strictEqual(blockHeight > 0, true);
  });

  test("initRpcClient prefers SOLANA_RPC_URL over the SOLANA_HOST default", async () => {
    process.env.SOLANA_HOST = "host.io";
    process.env.SOLANA_RPC_URL = "http://127.0.0.1:8899";

    try {
      const client = initRpcClient();
      const blockHeight = await client.getBlockHeight().send();
      assert.strictEqual(blockHeight > 0, true);
    } finally {
      delete process.env.SOLANA_HOST;
      delete process.env.SOLANA_RPC_URL;
    }
  });

  test("initRpcSubscriptionsClient", async () => {
    const client = initRpcSubscriptionsClient();
    const slotNotifications = await client.slotNotifications().subscribe({ abortSignal: AbortSignal.timeout(500) });

    for await (const notification of slotNotifications) {
      assert.strictEqual(notification.slot > 0, true);
    }
  });

  describe("waitForSlot", () => {
    test("successfully waits for a slot", async () => {
      const client = initRpcClient();
      const currentSlot = await client.getSlot({ commitment: "confirmed" }).send();
      const atSlot = await waitForSlot(currentSlot + 1n);
      assert.ok(atSlot);
    });

    test("slot is not reached within timeout", async () => {
      const client = initRpcClient();
      const currentSlot = await client.getSlot({ commitment: "confirmed" }).send();
      const atSlot = await waitForSlot(currentSlot + 50n, 500);
      assert.strictEqual(atSlot, false);
    });
  });
});

File: fragments/solana_rpc/solana_rpc_utils_test.go
package solana_rpc

import (
	"context"
	"testing"
	"time"

	"github.com/gagliardetto/solana-go/rpc"
)

func TestSolanaInitRpcClient(t *testing.T) {
	client := InitRpcClient()

	blockHeight, err := client.GetBlockHeight(context.Background(), rpc.CommitmentConfirmed)

	if err != nil {
		t.Errorf("Expected no error, but got %s", err)
	}

	if blockHeight < 1 {
		t.Errorf("Expected block height to be greater than 0, but got %d", blockHeight)
	}
}

func TestSolanaInitRpcSubscriptionsClient(t *testing.T) {
	client := InitRpcSubscriptionsClient()

	sub, err := client.SlotSubscribe()
	if err != nil {
		t.Fatalf("Failed to subscribe to slots: %v", err)
	}
	defer sub.Unsubscribe()

	ctxWithTimeout, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	slot, err := sub.Recv(ctxWithTimeout)

	if err != nil {
		t.Fatalf("Failed to receive slot update: %v", err)
	}

	if slot.Slot < 1 {
		t.Errorf("Expected slot to be greater than 0, but got %d", slot.Slot)
	}
}

func TestSolanaWaitForSlot_SlotReached(t *testing.T) {
	client := InitRpcClient()
	recentSlot, err := client.GetSlot(context.Background(), rpc.CommitmentConfirmed)
	if err != nil {
		t.Fatalf("GetSlot failed: %v", err)
	}

	success, err := WaitForSlot(recentSlot+1, nil)
	if err != nil {
		t.Errorf("Expected no error, but got %s", err)
	}
	if !success {
		t.Errorf("Expected success to be true, but got false")
	}
}

func TestSolanaWaitForSlot_Timeout(t *testing.T) {
	client := InitRpcClient()
	recentSlot, err := client.GetSlot(context.Background(), rpc.CommitmentConfirmed)
	if err != nil {
		t.Fatalf("GetSlot failed: %v", err)
	}

	timeout := uint64(10)
	success, err := WaitForSlot(recentSlot+50, &timeout)
	if err != nil {
		t.Errorf("Expected no error, but got %s", err)
	}
	if success {
		t.Errorf("Expected failure, but it passed")
	}
}
